lnst/Recipes/ENRT/MeasurementGenerators/BaseFlowMeasurementGenerator.py

```diff
@@ -101,66 +104,64 @@ def generate_flow_combinations(self, config):
         :any:`msg_sizes`.
 
         :return: list of Flow combinations to measure in parallel
-        :rtype: List[:any:`PerfFlow`]
+        :rtype: Iterator[:any:`PerfFlow`]
         """
-        for client_nic, server_nic in self.generate_perf_endpoints(config):
-            for ipv in self.params.ip_versions:
-                for perf_test in self.params.perf_tests:
-                    for size in self.params.perf_msg_sizes:
-                        yield self._create_perf_flows(
-                            perf_test,
-                            ipv,
-                            client_nic,
-                            server_nic,
-                            size,
-                        )
-
-    def generate_perf_endpoints(self, config):
-        """Generator for perf endpoints
-
-        To be overridden by a derived class.
-
-        :return: list of device pairs
-        :rtype: List[Tuple[:any:`Device`, :any:`Device`]]
+        def is_wanted_ip_version(endpoint_pair: EndpointPair[IPEndpointLike]) -> bool:
+            ipv_string = "ipv4" if isinstance(endpoint_pair.first.address, Ip4Address) else "ipv6"
+            return ipv_string in self.params.ip_versions
+
+        for parallel_endpoint_pairs in self.generate_perf_endpoints(config):
+            filtered_parallel_endpoints = [
+                endpoint_pair
+                for endpoint_pair in parallel_endpoint_pairs
+                if is_wanted_ip_version(endpoint_pair)
+            ]
+
+            # skip test if no endpoints left after ip filtering
+            if not filtered_parallel_endpoints:
+                continue
+
+            for perf_test in self.params.perf_tests:
+                for size in self.params.perf_msg_sizes:
+                    yield self._create_perf_flows(
+                        filtered_parallel_endpoints,
+                        perf_test,
+                        size,
+                    )
+
+    def generate_perf_endpoints(self, config: EnrtConfiguration) -> list[Collection[EndpointPair[IPEndpointLike]]]:
+        """
+        Returns sequential list of parallel endpoint-pairs that should be performance-tested.
         """
         return []
 
     def _create_perf_flows(
         self,
-        perf_test,
-        ipv,
-        client_nic,
-        server_nic,
+        endpoint_pairs: list[EndpointPair[IPEndpoint]],
+        perf_test: str,
         msg_size,
-    ) -> List[PerfFlow]:
-
-        ip_filter = {}
-        if ipv == "ipv4":
-            ip_filter.update(family=AF_INET)
-        elif ipv == "ipv6":
-            ip_filter.update(family=AF_INET6)
-            ip_filter.update(is_link_local=False)
-
-        client_ips = client_nic.ips_filter(**ip_filter)
-        server_ips = server_nic.ips_filter(**ip_filter)
+    ) -> list[PerfFlow]:
+        port_iter = itertools.count(12000)
```

Comment:
I think we should initialize the iterator for each endpoint pair. So, move it before:
```
for i in range(self.params.perf_parallel_processes):
```

Reply:
The generated `PerfFlow` instances are ran in parallel (see description of return value of `generate_flow_combinations()`. Therefore I think these can't have the same ports.

Reply:
Typically the enpoints have different IP addresses, so using the same port should be ok. Btw, are there endpoints that can define the ports themselves. If so, the generator should use them.

Reply:
having a different port number should be fine regardless of if the ip addresses of the endpoints are the same or different, the space for port numbers is big enough that we likely can't exhaust it with a test like this....

on the other hand if the endpoint ips are the same then having the same port number would be an issue...

i think the decision for this is what keeps the closes backwards compatibility - if we apply this version, will the tests that utilize multiple parallel flows be all different or will they stay the same? Does this change anything wrt. looking for baselines in catalog?

---

i aggree with the "if an endpoint defines a port it should be used" instead of overwriting it with a new value though...

Reply:
> if we apply this version, will the tests that utilize multiple parallel flows be all different or will they stay the same?

I think everything stays the same until some recipe defines multiple IPs of the same type (let's say IPv4) on both endpoints defined in `generate_perf_endpoints()`. Currently, we do a product on all defined IPs of the same type and run these in *parallel*. See the definition of `ip_endpoints`. Maybe we should change that to be sequential?

Previously if A and B each defined multiple IPs, only the first pair would be picked I think. Currently, we pick all of them *at once*.
```
previously:
A.a <-> B.a

now:
A.a <-> B.a
A.a <-> B.b
A.b <-> B.a
A.b <-> B.b

Reply:
> I think everything stays the same until some recipe defines multiple IPs of the same type

Yes, I think this should not be an issue. I'm fine with different ports based on your arguments.

Reply:
> i aggree with the "if an endpoint defines a port it should be used" instead of overwriting it with a new value though...

the question is how this should work if there are multiple processes. because there would be a conflict if the same port is reused for all measurements.

Reply:
> > i aggree with the "if an endpoint defines a port it should be used" instead of overwriting it with a new value though...
> 
> the question is how this should work if there are multiple processes. because there would be a conflict if the same port is reused for all measurements.

IMO.... if the tester requests specific ports with specific endpoints... then these should be used exactly as requested... if this doesn't work because there is a port conflict... that's either intended and expected by the tester... or it's a bug caused by the tester and should be fixed... we shouldn't attempt to fix his or disallow these scenarios from the tester.

----

> Previously if A and B each defined multiple IPs, only the first pair would be picked I think. Currently, we pick all of them at once.

i'm ok with this, especially since we don't have a test configuration that would be affected by this at this moment. The behaviour imo makes sense, I just want to make sure we're not unknowingly changing something that previously worked one way to work another way without proper checks...

Reply:
So... the `generate_perf_endpoints()` returns a list of lists of `EndpointPair[IPEndpointLike]`. You basically want to make sure that if someone _does_ return `IPPortEndpoint` instead of `IPEndpoint`, you want to respect the port. I don't think adding a `isinstance` check to `generate_flow_combinations()` where the ports are used is a good idea.

Maybe we should change `generate_perf_endpoints()` return value to be `IPPortEndpoint` and let the recipe author pick the ports themselves. I believe this is the cleanest solution. Do you agree? @olichtne @jtluka

Reply:
that would mean that the logic of "generate 100 parallel flows" would move into the specific recipes where the `generate_perf_endpoints` would have to be the generator that integrates the port number generation correct?

it sounds... correct, but i think that means that it also will generate a bunch of duplication that we will want somehow move into a shared/common function or at least a helper....

soooo... does that just end up being moved to the same level where the current `generate_flow_combinations` method is? because if so... we're back to the current generator solution?

Reply:
Hmm... right, I forgot about the `perf_parallel_processes` param... What should we do in a scenario where a tester requests `perf_parallel_processes=1000`, but also returns endpoints with specific ports? We can't follow the ports then. Do we just error out?

Reply:
This would mean we *will* need to `isinstance` check the returned endpoint pairs... Or we make `IPPortEndpoint` class have `port: Optional[int]`?

Personally, I would _not_ support custom port specification _for now_. Currently, we do not need it. I would like to finally merge this. We can leave this for a different PR.

Reply:
yeah i think we can postpone this. open an issue so we have a reminder about it. @jtluka  do  you agree?

Reply:
Sure, let's followup on this in separate issue.

However, I want to make sure that the parallel iperf test scenarios work fine after this change. Looking at the results of the linked job it seems it's not.

Original:
```
2023-08-31 10:36:18       (localhost)        -    INFO: 
    CPU Baseline average evaluation for Host host1:
    PASS: CPU cpu: 1514_utilization 6.87% lower than baseline. Allowed difference: 7.94%
2023-08-31 10:36:18       (localhost)        -    INFO: Result: PASS, What:
2023-08-31 10:36:18       (localhost)        -    INFO: 
    CPU Baseline average evaluation for Host host2:
    PASS: CPU cpu: 1515_utilization 1.75% lower than baseline. Allowed difference: 4.71%
```

New:
```
2023-09-04 19:49:26       (localhost)        -    INFO: 
    CPU Baseline average evaluation for Host host1:
    FAIL: CPU cpu: 1512_utilization no threshold found
2023-09-04 19:49:26       (localhost)        -    INFO: Result: FAIL, What:
2023-09-04 19:49:26       (localhost)        -    INFO: 
    CPU Baseline average evaluation for Host host2:
    FAIL: CPU cpu: 1513_utilization no threshold found
```

Seems that indexes got broken.

Reply:
It looks like that two calls to `ips_filter()` were removed:
```
2023-08-31 10:29:57       (localhost)        -    INFO: Result: PASS, What: Calling Device method host1.eth0.ips_filter(family=<AddressFamily.AF_INET: 2>)
2023-08-31 10:29:57       (localhost)        -    INFO: Result: PASS, What: Calling Device method host2.eth0.ips_filter(family=<AddressFamily.AF_INET: 2>)
```

I assume this is expected, just mentioning here so that we can prepare for downstream changes. I will check if this impacts single stream (basically all tests).

Reply:
Single stream are also impacted. I'll close this thread as it's unrelated to the original issue discussed here.